File: idasdk61/plugins/formchooser/formchooser.cpp
/*
 *  This plugin demonstrates how to use choosers inside forms.
 *
 */

#include <ida.hpp>
#include <idp.hpp>
#include <loader.hpp>
#include <kernwin.hpp>

//--------------------------------------------------------------------------
// raw data of the png icon (16x16)
static unsigned char icon_data[182] =
{
  0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A, 0x00, 0x00, 0x00, 0x0D, 0x49, 0x48, 0x44, 0x52,
  0x00, 0x00, 0x00, 0x10, 0x00, 0x00, 0x00, 0x10, 0x08, 0x06, 0x00, 0x00, 0x00, 0x1F, 0xF3, 0xFF,
  0x61, 0x00, 0x00, 0x00, 0x7D, 0x49, 0x44, 0x41, 0x54, 0x78, 0xDA, 0x63, 0x64, 0xC0, 0x0E, 0xFE,
  0xE3, 0x10, 0x67, 0x24, 0x28, 0x00, 0xD2, 0xFC, 0xF3, 0xAF, 0x36, 0x56, 0xDD, 0xEC, 0xCC, 0x57,
  0x31, 0xF4, 0x20, 0x73, 0xC0, 0xB6, 0xE2, 0xD2, 0x8C, 0x66, 0x08, 0x5C, 0x2F, 0x8A, 0x01, 0x84,
  0x34, 0x63, 0x73, 0x09, 0x23, 0xA9, 0x9A, 0xD1, 0x0D, 0x61, 0x44, 0xD7, 0xCC, 0xCF, 0x02, 0x71,
  0xE2, 0xC7, 0x3F, 0xA8, 0x06, 0x62, 0x13, 0x07, 0x19, 0x42, 0x7D, 0x03, 0x48, 0xF5, 0xC6, 0x20,
  0x34, 0x00, 0xE4, 0x57, 0x74, 0xFF, 0xE3, 0x92, 0x83, 0x19, 0xC0, 0x40, 0x8C, 0x21, 0xD8, 0x34,
  0x33, 0x40, 0xA3, 0x91, 0x01, 0x97, 0x21, 0xC8, 0x00, 0x9B, 0x66, 0x38, 0x01, 0x33, 0x00, 0x44,
  0x50, 0x92, 0x94, 0xB1, 0xBA, 0x04, 0x8B, 0x66, 0x9C, 0x99, 0x09, 0xC5, 0x10, 0x1C, 0xE2, 0x18,
  0xEA, 0x01, 0xA3, 0x65, 0x55, 0x0B, 0x33, 0x14, 0x07, 0x63, 0x00, 0x00, 0x00, 0x00, 0x49, 0x45,
  0x4E, 0x44, 0xAE, 0x42, 0x60, 0x82
};
static int icon_id = 0;

//--------------------------------------------------------------------------
// column widths
static const int widths[] = { 40 };

// column headers
static const char *header[] =
{
  "Item",
};
CASSERT(qnumber(widths) == qnumber(header));

//--------------------------------------------------------------------------
static int main_current_index = 1;

//-------------------------------------------------------------------------
// function that generates the list line
static void idaapi aux_choose_getl(void * /*obj*/, uint32 n, char *const *arrptr)
{
  if ( n == 0 ) // generate the column headers
    qstrncpy(arrptr[0], header[0], MAXSTR);
  else
    qsnprintf(arrptr[0], MAXSTR, "Item %d", n);
}

//-------------------------------------------------------------------------
// function that returns number of lines in the list
static uint32 idaapi main_choose_sizer(void * /*obj*/)
{
  return 10;
}

//-------------------------------------------------------------------------
// function that returns number of lines in the list
static uint32 idaapi aux_choose_sizer(void * /*obj*/)
{
  return main_current_index;
}

//-------------------------------------------------------------------------
// function that generates the list line
static void idaapi main_choose_getl(void * /*obj*/,uint32 n, char *const *arrptr)
{
  // generate the column headers
  if ( n == 0 )
    qstrncpy(arrptr[0], header[0], MAXSTR);
  else
    qsnprintf(arrptr[0], MAXSTR, "Option %d", n);
}

//-------------------------------------------------------------------------
static void refresh_selection_edit(form_actions_t & fa)
{
  static char str[MAXSTR], tmp[100];
  static intvec_t array;

  if ( main_current_index == 0 )
    qstrncpy(str, "No selection", sizeof(str));
  else
  {
    qsnprintf(str, sizeof(str), "Main %d", main_current_index);

    fa.get_field_value(4, &array);
    if ( array.size() > 0 )
    {
      qstrncat(str, " - Aux item(s) ", sizeof(str));
      for ( int i = 0; i < array.size() - 1; i++ )
      {
        qsnprintf(tmp, sizeof(tmp), "%d, ", array.at(i));
        qstrncat(str, tmp, sizeof(str));
      }
      qsnprintf(tmp, sizeof(tmp), "%d", array.at(array.size() - 1));
      qstrncat(str, tmp, sizeof(str));
    }
  }

  fa.set_field_value(5, str);
}

//---------------------------------------------------------------------------
static uint32 idaapi chooser_menu_cmd(void * /*obj*/, uint32 /*n*/)
{
  msg("Menu item clicked!\n");
  return 1;
}

//--------------------------------------------------------------------------
static int idaapi modcb(int fid, form_actions_t &fa)
{
  static intvec_t array;
  switch ( fid )
  {
    case -1:
      msg("initializing\n");
      refresh_selection_edit(fa);
      // add an item to the context menu of the main chooser
      // note: the caption "Form with choosers:3" contains the title of the
      //       form followed by a colon delimiter and the field id of the
      //       embedded chooser
      add_chooser_command("Form with choosers:3", "Test", chooser_menu_cmd,
                          -1, icon_id, CHOOSER_POPUP_MENU);
      break;
    case -2:
      msg("terminating\n");
      break;
    // main chooser
    case 3:
      msg("main chooser selection change\n");
      fa.get_field_value(3, &array);
      if ( array.size() > 0 )
        main_current_index = array[0];
      else
        main_current_index = 0;
      // refresh auxiliar chooser
      fa.refresh_field(4);
      refresh_selection_edit(fa);
      break;
    // auxiliar chooser
    case 4:
      refresh_selection_edit(fa);
      break;
    // Aux value text control
    case 5:
      break;
    default:
      msg("unknown id %d\n", fid);
      break;
  }

  return 1;
}

//--------------------------------------------------------------------------
static void idaapi run(int)
{
  static const char form[] =
    "STARTITEM 0\n"
    "Form with choosers\n\n"
    "%/"
    "Select an item in the main chooser:\n"
    "\n"
    "<Main chooser:E3::30::><Auxiliar chooser (multi):E4::30::>\n\n"
    "<Selection:A5:1023:40::>\n"
    "\n";

  static chooser_info_t main_chi =
  {
    sizeof(chooser_info_t),
    CH_NOIDB, // flags (doesn't need an open database)
    0, 0, // width, height
    NULL, //title
    NULL, // obj
    qnumber(header), // columns
    widths,
    icon_id, // icon
    0, // deflt
    NULL, // popup_names
    main_choose_sizer,
    main_choose_getl,
    NULL, // del
    NULL, // ins
    NULL, // update
    NULL, // edit
    NULL, // enter
    NULL, // destroyer
    NULL, // get_icon
    NULL, // select
    NULL // refresh
  };

  chooser_info_t aux_chi = main_chi;
  aux_chi.flags |= CH_MULTI;
  aux_chi.sizer = aux_choose_sizer;
  aux_chi.getl = aux_choose_getl;

  intvec_t main_sel, aux_sel;
  char str[MAXSTR];
  str[0] = '\0';

  // default selection for the main chooser
  main_sel.push_back(main_current_index);

  if ( AskUsingForm_c(form, modcb,
                      &main_chi, &main_sel,
                      &aux_chi, &aux_sel,
                      str) > 0 )
  {
    msg("Selection: %s\n", str);
  }
}

//--------------------------------------------------------------------------
static int idaapi init(void)
{
  icon_id = load_custom_icon(icon_data, sizeof(icon_data), "png");
  return PLUGIN_OK;
}

//--------------------------------------------------------------------------
static void idaapi term(void)
{
  free_custom_icon(icon_id);
}

//--------------------------------------------------------------------------
plugin_t PLUGIN =
{
  IDP_INTERFACE_VERSION,
  PLUGIN_UNL,
  init,                  // initialize
  term,                  // terminate. this pointer may be NULL.
  run,                   // invoke plugin
  NULL,                  // long comment about the plugin
  NULL,                  // multiline help about the plugin
  "Forms chooser sample",// the preferred short name of the plugin
  NULL                   // the preferred hotkey to run the plugin
};

File: idasdk61/plugins/formchooser/formchooser.py
import idaapi
from idaapi import Choose2
from idaapi import Form

#<pycode(ex_formchooser)>
# --------------------------------------------------------------------------
class MainChooserClass(Choose2):
    def __init__(self, title, icon):
        Choose2.__init__(self,
                         title,
                         [ ["Item", 10] ],
                         icon=icon,
                         flags=Choose2.CH_NOIDB,
                         embedded=True, width=30, height=20)

    def OnClose(self):
        pass

    def OnGetLine(self, n):
        return ["Option %d" % n]

    def OnGetSize(self):
        return 10

    def OnCommand(self, n, cmd_id):
        if cmd_id == self.cmd_id1:
            print("Context menu on: %d" % n)

        return 0


# --------------------------------------------------------------------------
class AuxChooserClass(Choose2):
    def __init__(self, title, icon):
        Choose2.__init__(self,
                         title,
                         [ ["Item", 10] ],
                         icon=icon,
                         flags=Choose2.CH_NOIDB | Choose2.CH_MULTI,
                         embedded=True, width=30, height=20)

    def OnClose(self):
        pass

    def OnGetLine(self, n):
        return ["Item %d" % n]

    def OnGetSize(self):
        t = self.form.main_current_index
        return 0 if t < 0 else t+1


# --------------------------------------------------------------------------
class MyChooserForm(Form):

    # Custom icon data
    icon_data = (
        "\x89\x50\x4E\x47\x0D\x0A\x1A\x0A\x00\x00\x00\x0D\x49\x48\x44\x52"
        "\x00\x00\x00\x10\x00\x00\x00\x10\x08\x06\x00\x00\x00\x1F\xF3\xFF"
        "\x61\x00\x00\x00\x7D\x49\x44\x41\x54\x78\xDA\x63\x64\xC0\x0E\xFE"
        "\xE3\x10\x67\x24\x28\x00\xD2\xFC\xF3\xAF\x36\x56\xDD\xEC\xCC\x57"
        "\x31\xF4\x20\x73\xC0\xB6\xE2\xD2\x8C\x66\x08\x5C\x2F\x8A\x01\x84"
        "\x34\x63\x73\x09\x23\xA9\x9A\xD1\x0D\x61\x44\xD7\xCC\xCF\x02\x71"
        "\xE2\xC7\x3F\xA8\x06\x62\x13\x07\x19\x42\x7D\x03\x48\xF5\xC6\x20"
        "\x34\x00\xE4\x57\x74\xFF\xE3\x92\x83\x19\xC0\x40\x8C\x21\xD8\x34"
        "\x33\x40\xA3\x91\x01\x97\x21\xC8\x00\x9B\x66\x38\x01\x33\x00\x44"
        "\x50\x92\x94\xB1\xBA\x04\x8B\x66\x9C\x99\x09\xC5\x10\x1C\xE2\x18"
        "\xEA\x01\xA3\x65\x55\x0B\x33\x14\x07\x63\x00\x00\x00\x00\x49\x45"
        "\x4E\x44\xAE\x42\x60\x82")


    def Free(self):
        del self.EChMain.form
        del self.EChAux.form

        # Call base
        Form.Free(self)

        # Free icon
        if self.icon_id != 0:
            idaapi.free_custom_icon(self.icon_id)
            self.icon_id = 0


    def __init__(self):
        # Load custom icon
        self.icon_id = idaapi.load_custom_icon(data=MyChooserForm.icon_data)
        if self.icon_id == 0:
            raise RuntimeError("Failed to load icon data!")

        self.main_current_index = -1
        self.EChMain = MainChooserClass("MainChooser", self.icon_id)
        self.EChAux  = AuxChooserClass("AuxChooser", self.icon_id)

        # Link the form to the EChooser
        self.EChMain.form = self
        self.EChAux.form = self

        Form.__init__(self, r"""STARTITEM 0
Form with choosers

    {FormChangeCb}
    Select an item in the main chooser:

    <Main chooser:{ctrlMainChooser}><Auxiliar chooser (multi):{ctrlAuxChooser}>


    <Selection:{ctrlSelectionEdit}>

""", {
            'ctrlSelectionEdit' : Form.StringInput(),
            'FormChangeCb'      : Form.FormChangeCb(self.OnFormChange),
            'ctrlMainChooser'   : Form.EmbeddedChooserControl(self.EChMain),
            'ctrlAuxChooser'    : Form.EmbeddedChooserControl(self.EChAux),
        })


    def refresh_selection_edit(self):
        if self.main_current_index < 0:
            s = "No selection in the main chooser"
        else:
            s = "Main %d" % self.main_current_index

            # Get selection in the aux chooser
            sel = self.GetControlValue(self.ctrlAuxChooser)
            if sel:
                s = "%s - Aux item(s): %s" % (s, ",".join(str(x) for x in sel))

        # Update string input
        self.SetControlValue(self.ctrlSelectionEdit, s)


    def OnFormChange(self, fid):
        if fid == -1:
            print("Initialization")
            self.refresh_selection_edit()

            # Add an item to the context menu of the main chooser
            id = self.ctrlMainChooser.AddCommand("Test", icon=self.icon_id)
            print "id=%d" % id
            if id < 0:
                print("Failed to install menu for main embedded chooser")
            else:
                self.EChMain.cmd_id1 = id

        elif fid == -2:
            print("Terminating");

        elif fid == self.ctrlMainChooser.id:
            print("main chooser selection change");
            l = self.GetControlValue(self.ctrlMainChooser);
            if not l:
                self.main_current_index = -1
            else:
                self.main_current_index = l[0]

            # Refresh auxiliar chooser
            self.RefreshField(self.ctrlAuxChooser)
            self.refresh_selection_edit()

        elif fid == self.ctrlAuxChooser.id:
            self.refresh_selection_edit()

        elif fid == self.ctrlSelectionEdit.id:
            pass
        else:
            print("unknown id %d" % fid)

        return 1

#</pycode(ex_formchooser)>

def main():
    global f
    f = MyChooserForm()
    try:
        f.Compile()
        r = f.Execute()
        print("Execute returned: %d" % r)
        f.Free()
    except Exception as e:
        print("Failed to show form: %s" % str(e))

if __name__=='__main__':
    main()